coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/UnionSqlCallTransformer.java

```diff
@@ -0,0 +1,165 @@
+ *
+ * @see <a href="https://github.com/trinodb/trino/issues/9031">Change char/varchar coercion in Trino</a>
+ */
+public class UnionSqlCallTransformer extends SqlCallTransformer {
+
+  public UnionSqlCallTransformer(TypeDerivationUtil typeDerivationUtil) {
+    super(typeDerivationUtil);
+  }
+
+  @Override
+  protected boolean condition(SqlCall sqlCall) {
+    return sqlCall.getOperator().kind == SqlKind.UNION || sqlCall.getOperator().kind == SqlKind.INTERSECT
+        || sqlCall.getOperator().kind == SqlKind.MINUS;
+  }
+
+  @Override
+  protected SqlCall transform(SqlCall sqlCall) {
+    List<SqlNode> operandsList = sqlCall.getOperandList();
+    List<List<Optional<RelDataType>>> columnsTypesLists = new ArrayList<>();
+    for (SqlNode operand : operandsList) {
+      if (operand.getKind() == SqlKind.SELECT && ((SqlSelect) operand).getSelectList() != null) {
+        SqlSelect select = (SqlSelect) operand;
+        List<SqlNode> selectNodes = select.getSelectList().getList();
+        if (columnsTypesLists.isEmpty()) {
+          for (int i = 0; i < selectNodes.size(); i++) {
+            columnsTypesLists.add(new ArrayList<>());
+          }
+        }
+
+        for (int i = 0; i < selectNodes.size(); i++) {
+          SqlNode sqlNode = selectNodes.get(i);
+          if (sqlNode.getKind() == SqlKind.IDENTIFIER && ((SqlIdentifier) sqlNode).isStar()) {
+            // The type derivation of * on the SqlNode layer is not supported now.
+            return sqlCall;
+          }
+
+          Optional<RelDataType> selectNodeDataType;
+          try {
+            selectNodeDataType = Optional.of(deriveRelDatatype(sqlNode));
+          } catch (RuntimeException e) {
+            // The type derivation may fail for complex expressions
+            selectNodeDataType = Optional.empty();
+          }
+          columnsTypesLists.get(i).add(selectNodeDataType);
+        }
+      } else {
+        return sqlCall;
+      }
+    }
+
+    List<SqlNode> updatedOperands = new ArrayList<>(sqlCall.getOperandList().size());
+    for (SqlNode operand : operandsList) {
+      SqlSelect select = (SqlSelect) operand;
+
+      List<SqlNode> selectNodes = select.getSelectList().getList();
+
+      List<SqlNode> rewrittenSelectNodes = new ArrayList<>();
+      boolean useRewrittenSelectNodes = false;
+      for (int i = 0; i < selectNodes.size(); i++) {
+
+        if (columnsTypesLists.get(i).stream().anyMatch(columnType -> !columnType.isPresent())) {
+          // Couldn't determine the type for all the expressions
+          continue;
+        }
+        RelDataType inferredColumnType =
+            leastRestrictive(columnsTypesLists.get(i).stream().map(Optional::get).collect(Collectors.toList()));
+
+        SqlNode selectNode = selectNodes.get(i);
+        RelDataType selectNodeDataType = deriveRelDatatype(selectNodes.get(i));
+        if (!selectNodeDataType.equals(inferredColumnType) && selectNodeDataType.getSqlTypeName() == SqlTypeName.CHAR
```

Comment:
is `!selectNodeDataType.equals(inferredColumnType)` redundant? The 2nd and 3rd conditions seem to imply that.